File: web/src/lib/api-client.ts
import { ApiError } from './api-error';
import { env } from './env';
import type { AuthTokens } from '@/types/api';

type TokenGetter = () => string | null;
type TokenSetter = (tokens: AuthTokens) => void;
type SessionExpiredHandler = () => void;

interface RequestOptions extends Omit<RequestInit, 'body'> {
  body?: unknown;
  /** Skip the Authorization header (login, refresh, public careers page). */
  anonymous?: boolean;
  query?: Record<string, string | number | boolean | string[] | undefined | null>;
}

/**
 * Single HTTP entry point.
 *
 * Handles one thing the rest of the app should never think about: when the
 * access token expires mid-session, the first 401 triggers a refresh and every
 * concurrent request waits on that single refresh rather than stampeding the
 * endpoint and invalidating each other's rotated tokens.
 */
class ApiClient {
  private getAccessToken: TokenGetter = () => null;
  private getRefreshToken: TokenGetter = () => null;
  private onTokensRefreshed: TokenSetter = () => {};
  private onSessionExpired: SessionExpiredHandler = () => {};
  private refreshInFlight: Promise<string | null> | null = null;

  configure(handlers: {
    getAccessToken: TokenGetter;
    getRefreshToken: TokenGetter;
    onTokensRefreshed: TokenSetter;
    onSessionExpired: SessionExpiredHandler;
  }): void {
    this.getAccessToken = handlers.getAccessToken;
    this.getRefreshToken = handlers.getRefreshToken;
    this.onTokensRefreshed = handlers.onTokensRefreshed;
    this.onSessionExpired = handlers.onSessionExpired;
  }

  get<T>(path: string, options: RequestOptions = {}): Promise<T> {
    return this.request<T>('GET', path, options);
  }

  post<T>(path: string, body?: unknown, options: RequestOptions = {}): Promise<T> {
    return this.request<T>('POST', path, { ...options, body });
  }

  patch<T>(path: string, body?: unknown, options: RequestOptions = {}): Promise<T> {
    return this.request<T>('PATCH', path, { ...options, body });
  }

  put<T>(path: string, body?: unknown, options: RequestOptions = {}): Promise<T> {
    return this.request<T>('PUT', path, { ...options, body });
  }

  delete<T>(path: string, options: RequestOptions = {}): Promise<T> {
    return this.request<T>('DELETE', path, options);
  }

  private async request<T>(method: string, path: string, options: RequestOptions): Promise<T> {
    const url = env.apiBaseUrl + path + buildQueryString(options.query);
    const response = await this.send(method, url, options);

    if (response.status === 401 && !options.anonymous) {
      const token = await this.refreshOnce();
      if (!token) {
        this.onSessionExpired();
        throw ApiError.fromBody(401, await safeJson(response));
      }
      const retried = await this.send(method, url, options);
      return this.parse<T>(retried);
    }

    return this.parse<T>(response);
  }

  private send(method: string, url: string, options: RequestOptions): Promise<Response> {
    const headers = new Headers(options.headers);
    headers.set('Accept', 'application/json');

    let body: BodyInit | undefined;
    if (options.body !== undefined) {
      if (options.body instanceof FormData) {
        body = options.body;
      } else {
        headers.set('Content-Type', 'application/json');
        body = JSON.stringify(options.body);
      }
    }

    if (!options.anonymous) {
      const token = this.getAccessToken();
      if (token) headers.set('Authorization', `Bearer ${token}`);
    }

    return fetch(url, { ...options, method, headers, body });
  }

  private async parse<T>(response: Response): Promise<T> {
    if (response.status === 204) return undefined as T;

    if (!response.ok) {
      throw ApiError.fromBody(response.status, await safeJson(response));
    }

    const text = await response.text();
    return (text ? JSON.parse(text) : undefined) as T;
  }

  /** Collapses concurrent refresh attempts into a single in-flight request. */
  private refreshOnce(): Promise<string | null> {
    if (this.refreshInFlight) return this.refreshInFlight;

    const refreshToken = this.getRefreshToken();
    if (!refreshToken) return Promise.resolve(null);

    this.refreshInFlight = (async () => {
      try {
        const response = await fetch(`${env.apiBaseUrl}/auth/refresh`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ refreshToken }),
        });
        if (!response.ok) return null;

        const tokens = (await response.json()) as AuthTokens;
        this.onTokensRefreshed(tokens);
        return tokens.accessToken;
      } catch {
        return null;
      } finally {
        this.refreshInFlight = null;
      }
    })();

    return this.refreshInFlight;
  }
}

function buildQueryString(
  query: RequestOptions['query'],
): string {
  if (!query) return '';
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(query)) {
    if (value === undefined || value === null || value === '') continue;
    if (Array.isArray(value)) {
      if (value.length > 0) params.set(key, value.join(','));
    } else {
      params.set(key, String(value));
    }
  }
  const qs = params.toString();
  return qs ? `?${qs}` : '';
}

async function safeJson(response: Response): Promise<Record<string, unknown> | null> {
  try {
    return (await response.clone().json()) as Record<string, unknown>;
  } catch {
    return null;
  }
}

export const api = new ApiClient();

File: mobile/lib/core/network/api_client.dart
import 'dart:async';

import 'package:dio/dio.dart';

import '../config/app_config.dart';
import '../storage/token_storage.dart';
import 'api_exception.dart';

/// Called when the refresh token is rejected and the session is unrecoverable.
typedef SessionExpiredCallback = void Function();

/// HTTP client for the HRIS API.
///
/// Owns one thing the rest of the app should never think about: when the access
/// token expires mid-session, the first 401 triggers a refresh and every
/// concurrent request waits on that single refresh. Without this, a screen that
/// fires three requests at once would rotate the refresh token three times and
/// the server's reuse detection would revoke the whole session.
class ApiClient {
  ApiClient({
    required TokenStorage tokenStorage,
    required SessionExpiredCallback onSessionExpired,
    Dio? dio,
  })  : _tokenStorage = tokenStorage,
        _onSessionExpired = onSessionExpired,
        _dio = dio ??
            Dio(
              BaseOptions(
                baseUrl: AppConfig.apiBaseUrl,
                connectTimeout: AppConfig.connectTimeout,
                receiveTimeout: AppConfig.receiveTimeout,
                contentType: 'application/json',
                // Let the interceptor decide: Dio must not throw before we can
                // inspect a 401 and attempt a refresh.
                validateStatus: (int? status) => status != null && status < 500,
              ),
            ) {
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (RequestOptions options, RequestInterceptorHandler handler) async {
          if (options.extra['anonymous'] != true) {
            final String? token = await _tokenStorage.readAccessToken();
            if (token != null) {
              options.headers['Authorization'] = 'Bearer $token';
            }
          }
          handler.next(options);
        },
      ),
    );
  }

  final Dio _dio;
  final TokenStorage _tokenStorage;
  final SessionExpiredCallback _onSessionExpired;

  Future<String?>? _refreshInFlight;

  Future<T> get<T>(
    String path, {
    Map<String, dynamic>? query,
    bool anonymous = false,
  }) =>
      _send<T>('GET', path, query: query, anonymous: anonymous);

  Future<T> post<T>(
    String path, {
    Object? body,
    Map<String, dynamic>? query,
    bool anonymous = false,
  }) =>
      _send<T>('POST', path, body: body, query: query, anonymous: anonymous);

  Future<T> patch<T>(String path, {Object? body}) => _send<T>('PATCH', path, body: body);

  Future<T> delete<T>(String path, {Object? body}) => _send<T>('DELETE', path, body: body);

  Future<T> _send<T>(
    String method,
    String path, {
    Object? body,
    Map<String, dynamic>? query,
    bool anonymous = false,
  }) async {
    Response<dynamic> response;
    try {
      response = await _dio.request<dynamic>(
        path,
        data: body,
        queryParameters: query,
        options: Options(method: method, extra: <String, dynamic>{'anonymous': anonymous}),
      );
    } on DioException catch (error) {
      if (error.type == DioExceptionType.connectionError ||
          error.type == DioExceptionType.connectionTimeout) {
        throw ApiException.offline;
      }
      throw ApiException.fromResponse(
        error.response?.statusCode ?? 500,
        error.response?.data,
      );
    }

    if (response.statusCode == 401 && !anonymous) {
      final String? refreshed = await _refreshOnce();
      if (refreshed == null) {
        _onSessionExpired();
        throw ApiException.fromResponse(401, response.data);
      }
      return _send<T>(method, path, body: body, query: query, anonymous: anonymous);
    }

    final int status = response.statusCode ?? 500;
    if (status >= 400) {
      throw ApiException.fromResponse(status, response.data);
    }

    return response.data as T;
  }

  /// Collapses concurrent refresh attempts into one in-flight request.
  Future<String?> _refreshOnce() {
    final Future<String?>? existing = _refreshInFlight;
    if (existing != null) return existing;

    final Future<String?> attempt = _performRefresh();
    _refreshInFlight = attempt;
    return attempt.whenComplete(() => _refreshInFlight = null);
  }

  Future<String?> _performRefresh() async {
    final String? refreshToken = await _tokenStorage.readRefreshToken();
    if (refreshToken == null) return null;

    try {
      final Response<dynamic> response = await _dio.post<dynamic>(
        '/auth/refresh',
        data: <String, dynamic>{'refreshToken': refreshToken},
        options: Options(extra: <String, dynamic>{'anonymous': true}),
      );

      if (response.statusCode != 200 || response.data is! Map) return null;

      final Map<dynamic, dynamic> data = response.data as Map<dynamic, dynamic>;
      final String? access = data['accessToken'] as String?;
      final String? refresh = data['refreshToken'] as String?;
      if (access == null || refresh == null) return null;

      await _tokenStorage.saveTokens(accessToken: access, refreshToken: refresh);
      return access;
    } on DioException {
      return null;
    }
  }
}
